### PR TITLE
Issue #2134:  Add support for "Relationships" to Layout Contexts

### LINKED ISSUE #2134: Add support for "Relationships" to Layout Contexts.
In many situations when working with Layouts you want or need information _related_ to the primary thing being displayed. For example if you had created a layout configuration for the path `node/%`, you have the "Node" context available but not the "User" context for the author of that content. This means that you can place the fields from the node as blocks, but not fields from the node author user account.

By adding relationship support, it would make it possible to display blocks (or use conditions) from the content being displayed but from any object in Backdrop, as long as it was related to the original piece of content in some way.

This has come up a few times in issues such as #1815 (Does it make sense to allow multiple contexts for a single layout?) plus a few more related questions (#1815) and in [the original Layout module issue #86](https://github.com/backdrop/backdrop-issues/issues/86#issuecomment-56255644) as an item that needs to be done eventually, but we hadn't yet had an issue for it.

**Update: How this all works, a summary:**
Blocks display things, and often they need to display things from entities. For example you might have a layout with a "Node information block" in that layout. The code for that block *could* check which path you're on, and have a bit like `if arg(1) == 'node' && arg(2) is a number etc etc then load that node with that NID` to get the node details, and that's fine. 

But Backdrop also arranges that a Layout can provide that info directly to the blocks code so the dev doesn't have to check if the block is on a node page. So your block code simply checks for `does this layout have a node` (something like that). 

Layout would *store* that node info in a "container" called a context.

So Layouts with "node paths" like `node/%` automatically provide a node context to its blocks. Layouts with "user paths" like `user/%` provide a user. We recently made it so that you can manually create a new context container on your layout, with the "custom context" functionality. 

But what if you want to show the author of the node on all node pages? There couldn't be a "user container" on that layout. 

This PR then allows you to create context containers storing context that are related to other contexts already on the layout. So if your layout already has a node context, you can tell layout to automatically create another context with the author of that node, by adding a "User from node" relationship.

When you visit a page using that layout, say at a `node/%` path, and you've added a "User from node" relationship, in that layout there will then be two containers holding contexts, one with the node, another with the author of that node. Blocks that want node contexts would show up, and blocks that want a user context would also be happy.

**Advocate**: @docwilmot 

---
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/1733~~
~~Replaced by PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3422~~
Replaced by https://github.com/backdrop/backdrop/pull/3611 by @docwilmot and @quicksketch

### PR DESCRIPTION
**Pull request by quicksketch**
Fixes https://github.com/backdrop/backdrop-issues/issues/2134

Builds on https://github.com/backdrop/backdrop/pull/3422

I'm filing this as a separate PR to get the full test suite running.

- Renames "user_from_node" relationship plugin to "author_from_node"
- Moves logic from `layout_get_context_from_relationships()` to a private method `Layout::getContextFromRelationships()`
- Builds in loading of context relationships directly into `Layout::getContexts()` so that relationships do not need to be loaded separately. They can just be requested to be loaded at the same time as all other relationships.

Making this last change required that we be able to differentiate between contexts that are added through relationships versus other mechanisms. This has been a problem with contexts from the beginning: You can't really tell where the context came from other than by a loose naming convention on the array keys (i.e. a integer comes from a menu path, "current_user" is always specially keyed, and now prefixing with "relationship_"). The way the context is added is now explicitly set in `LayoutContext::usageType`, a public property on the `LayoutContext` class.

### CODE REVIEW

**Comment by docwilmot** [comment]
@quicksketch its late here, so please go ahead and merge at will. Thanks for taking this on.

**Comment by herbdool** [comment]
I like the approach of turning it into a method. Looks good.

File: core/modules/layout/includes/layout.layout.inc [hunk 1]
```diff
@@ -76,6 +76,21 @@ function layout_layout_context_info() {
   return $info;
 }
 
+/**
+ * Implements hook_layout_relationship_info().
+ */
+function layout_layout_relationship_info() {
+  $info['author_from_node'] = array(
+    'title' => t('User from node'),
```
Comment (herbdool) [inline core/modules/layout/includes/layout.layout.inc]: I think this title should be updated too. Can we make it "Author of content"?

File: core/modules/layout/includes/layout.layout.inc [hunk 2]
```diff
@@ -76,6 +76,21 @@ function layout_layout_context_info() {
...
+    'class' => 'LayoutRelationshipAuthorFromNode',
+    'context' => 'node',
+    'context_label' => t('Node'),
+    'description' => t('Creates the author of a node as a user context.'),
```
Comment (herbdool) [inline core/modules/layout/includes/layout.layout.inc]: Also "author of content"

File: core/modules/layout/plugins/relationships/author_from_node.inc [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  /**
+   * Return a summary for this relationship.
+   */
+  function getAdminSummary() {
+    if (!empty($this->settings['context_parent'])) {
+      list($label, $value) = explode(':', $this->settings['context_parent']);
+      if ($label == 'position') {
+        return t('User from node in position @position', array('@position' => $value+1));
```
Comment (herbdool) [inline core/modules/layout/plugins/relationships/author_from_node.inc]: Should be "author" here and below.

**Review by herbdool — CHANGES_REQUESTED** [review]
Changes of user -> author

**Comment by quicksketch** [comment]
Thanks @herbdool, I made those string changes you suggested.

In testing I couldn't find a good reason to retrieve a list of contexts without the relationships included, except for the configuration UI where relationships are added, so to simplify the API I removed `getAllContexts()` and instead made returning all contexts, including relationships, the default.

Also I found that the broken handler code didn't work quite right (caused a few notices) so that is fixed now too.

File: core/modules/layout/includes/layout.layout.inc [hunk 4]
```diff
@@ -76,6 +76,21 @@ function layout_layout_context_info() {
...
 
+/**
+ * Implements hook_layout_relationship_info().
+ */
+function layout_layout_relationship_info() {
+  $info['author_from_node'] = array(
+    'title' => t('Author from content'),
+    'class' => 'LayoutRelationshipAuthorFromNode',
+    'context' => 'node',
+    'context_label' => t('Node'),
```
Comment (jenlampton) [inline core/modules/layout/includes/layout.layout.inc]: Shouldn't this also be `Content`?

File: core/modules/layout/includes/layout.layout.inc [hunk 5]
```diff
@@ -76,6 +76,21 @@ function layout_layout_context_info() {
...
+    'description' => t('Creates the author of a node as a user context.'),
```
Comment (jenlampton) [inline core/modules/layout/includes/layout.layout.inc]: `Creates the author of a node as a...` should probably also be `Creates the author of content as a...` or `Creates the author of a piece of content as a...`

File: core/modules/layout/layout.module [hunk 6]
```diff
@@ -339,6 +339,14 @@ function layout_menu() {
     'file' => 'layout.context.admin.inc',
   ) + $base;
 
+  $items['admin/structure/layouts/manage/%layout_tempstore/relationship/add'] = array(
+    'title' => 'Add context relationship',
```
Comment (jenlampton) [inline core/modules/layout/layout.module]: The label `Add relationship` is probably sufficient here.

File: core/modules/layout/plugins/relationships/author_from_node.inc [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    if (!empty($this->settings['context_parent'])) {
+      list($label, $value) = explode(':', $this->settings['context_parent']);
+      if ($label == 'position') {
+        return t('Author from content in position @position', array('@position' => $value+1));
+      }
+      else {
+        return t('Author from content @id', array('@id' => $value));
+      }
+    }
+    return t('Author from content');
```
Comment (jenlampton) [inline core/modules/layout/plugins/relationships/author_from_node.inc]: Maybe `Author of content` would sound more human?

**Comment by jenlampton** [comment]
Code review looks great, my only feedback are suggestions for user-facing text, which shouldn't block the merge. Doing some manual testing now :)

File: core/modules/layout/layout.module (hunk 6 above)
Comment (quicksketch) [reply core/modules/layout/layout.module]: Yeah, I agree. It's already under a second that says "Contexts", so the additional wording of "Context Relationship" is redundant.

File: core/modules/layout/includes/layout.layout.inc (hunk 4 above)
Comment (quicksketch) [reply core/modules/layout/includes/layout.layout.inc]: Existing Node contexts still show "Node", not "Content", so this is consistent with what we have at the moment.

**Pull request merged**